sympy/solvers/tests/test_ode.py

```diff
@@ -1103,11 +1104,31 @@ def test_solve_ics():
     assert solve_ics([Eq(f(x), C1*sin(x) + C2*cos(x))], [f(x)], [C1, C2],
         {f(0): 1, f(x).diff(x).subs(x, 0): 1}) == {C1: 1, C2: 1}
 
-    # XXX: Ought to be ValueError
-    raises(NotImplementedError, lambda: solve_ics([Eq(f(x), C1*sin(x) + C2*cos(x))], [f(x)], [C1, C2], {f(0): 1, f(pi): 1}))
+    assert solve_ics([Eq(f(x), C1*sin(x) + C2*cos(x))], [f(x)], [C1, C2],
+        {f(0): 1}) == {C2: 1}
+
+    # Some more complex tests
+    K, r, f0 = symbols('K r f0')
+    eq = Eq(f(x).diff(x), r * f(x) * (1 - f(x) / K))
+    assert (dsolve(eq, f(x), ics={f(0): f0})) == Eq(f(x), -K * f0 * exp(r * x) / (
+        (K - f0) * (-f0 * exp(r * x) / (K - f0) - 1)))
+    assert (dsolve(f(x).diff(x) * (f(x).diff(x, 2) - x), ics={f(0): 0})) \
+           == [Eq(f(x), 0), Eq(f(x), C2 * x + x ** 3 / 6)]
+    assert (dsolve(f(x).diff(x) * (f(x).diff(x, 2) - x), ics={f(0): 0, f(x).diff(x).subs(x, 1): 0})) \
+           == [Eq(f(x), 0), Eq(f(x), x ** 3 / 6 - x / 2)]
+
+    # Order dependent tests Refer to PR #16098
+    assert (dsolve(f(x).diff(x) * (f(x).diff(x, 2) - x), ics={f(0): 0, f(x).diff(x).subs(x, 0): 0})) \
+           == [Eq(f(x), 0), Eq(f(x), x ** 3 / 6)]
+    assert (dsolve(f(x).diff(x) * (f(x).diff(x, 2) - x), ics={f(x).diff(x).subs(x, 0): 0, f(0): 0})) \
+           == [Eq(f(x), 0), Eq(f(x), x ** 3 / 6)]
+
+    K, r, f0 = symbols('K r f0')
+    eq = Eq(f(x).diff(x), r * f(x) * (1 - f(x) / K))
+    assert(dsolve(eq, f(x), ics={f(0): f0})) == Eq(f(x), -K*f0*exp(r*x)/((K - f0)*(-f0*exp(r*x)/(K - f0) - 1)))
 
     # XXX: Ought to be ValueError
-    raises(ValueError, lambda: solve_ics([Eq(f(x), C1*sin(x) + C2*cos(x))], [f(x)], [C1, C2], {f(0): 1}))
```

Comment:
You've changed this from raising ValueError to raising NotImplementedError. ValueError would be better but actually I think this should really return an empty list like solve does when there are no solutions:
```
In [2]: solve([x-1, x+1], x)                                                                                                                                  
Out[2]: []
```
This is because we think we know that there are no solutions satisfying the initial conditions. I'm not sure about this though.